src/FreeAge/common/type_stats.hpp

```diff
@@ -0,0 +1,140 @@
+
+#include "FreeAge/common/damage.hpp"
+#include "FreeAge/common/free_age.hpp"
+#include "FreeAge/common/resources.hpp"
+
+static constexpr u32 typeStatsVersion = 1;
+
+enum class AttackType {
+  None,
+  /// TODO: rename ?
+  Default,
+  /// Full damage to all affected in the area of effect.
+  HomogeneousAreaOfEffect,
+  /// Full damage to primary target and 0.5 to all others in the area of effect.
+  TrampleHalf, // War Elephants, Battle Elephant
+  /// Full damage to primary target and 5 unaffected by armor damage in the area of effect.
+  TrampleFivePure, // Cataphracts, Druzhina
+  /// First projectile does full damage and other projectiles deal 3 pierce damage and 0 melee damage.
+  MultipleProjectiles, // Chu Ko Nu, Kipchak
+  
+  TownCenter, Castle, Tower, BombardTower
+};
+
+enum class GarrisonType {
+  None,
+  Units, // TransportShip ?
+  VillagersAndFootArchersAndInfantry, // TownCenter and towers
+  NonSiege, // Castle
+  Villagers, // Khmer houses
+  /// Units from production buildings can be trained directly into garrison,
+  /// but not garrison from outside.
+  Production,
+  /// Same as Production plus Relics can garrison
+  ProductionAndRelics
+};
+
+struct ObjectTypeStats {
+
+  float fireRate;
+  Damage damage;
+
+  float minRange;
+  float maxRange;
+
+  // Attack accuracy.
+  float accuracy;
+
+  AttackType attackType;
+
+  /// NOTE: Melee or Ranged is considered a sub type of each AttackType and can be calculated from projectileSpeed
+
+  /// The projectile speed or 0 if the attack is melee.
+  float projectileSpeed;
+
+  /// The area of effect of the attack or 0 if there is none.
+  float areaOfEffectSize;
+
+  /// The number of attacks performed. Special handling based on the attackType: 
+  /// - MultipleProjectiles: the number of projectiles
+  /// TODO: rename ?
+  int attacksCount;
+
+  /// Duration from start of attack until damage being applied to target represended by the ratio of the full attack
+  /// duration. To retrive the duration in seconds multiply by the attack duration.
+  /// TODO: rename ?
+  float attackDelay;
+
+  bool friendlyDamage;
+
+  /// Regenerated HP per minute (can be affected by technologies)
+  int regeneration;
+
+  int maxHp;
+
+  Armor armor;
+
+  /// The line of sight radius.
+  float lineOfSight;
+
+  /// The number of seconds needed for its creation.
+  float creationTime;
+  /// The resources needed for its creation.
+  ResourceAmount cost;
+
+  GarrisonType garrisonType;
+  int garrisonCapacity;
+
+  /// Based on the type:
+  /// - Villager, FishingShip: gather rate
+  /// - Production building: production speed
+  /// - Other: ?
+  float workRate;
+
+  /// TODO: doc
+  int conversionResistanceLevel;
+
+  // Used only by resource spot and animals
+  ResourceAmount resources;
+};
+
+enum class PopulationDemand {
+  None,
+  One,
+  Half
+};
+
+struct UnitTypeStats : public ObjectTypeStats {
+
+  /// The population space needed.
+  PopulationDemand populationDemand;
```

Comment:
Maybe this can be generalized in `ObjectTypeStats` such as to unify `UnitTypeStats::populationDemand` and `BuildingTypeStats::populationSpace`. The unified attribute could for example be positive if the object gives population space, and negative if the object takes up population space.

Buildings taking up population space need to be handled anyway due to the Feitoria building of the Portuguese, which takes up 20 population.

Reply:
This a good idea. How could we store the half population then? I really don't like having *2 everywhere, but it seems like the only solution... Any ideas?

Reply:
One idea would be making a `PopulationCount` class which has a descriptive interface, handles the *2 or /2 internally, and otherwise behaves like a number. It could have functions like `SetToOneHalf()`, `SetToInteger(int value)`, ...

Reply:
I knew you would have a the perfect solution! I will add this!